### PR TITLE
Update PlatformAgnostic case conversion functions to allow expanding-length strings

### LINKED ISSUE #421: Unicode: String.prototype.{toUpperCase,toLowerCase} don't handle characters that would increase the string length
For example, `'ß'.toUpperCase() === 'SS'` should be true, but Chakra doesn't handle that character and it's `'ß'.toUpperCase() === 'ß'` instead.

The list of such characters are maintained in the [SpecialCasing.txt](http://unicode.org/Public/UNIDATA/SpecialCasing.txt) file.

### LINKED ISSUE #4526: ChakraCore crashes with assertion failure in UnicodeText.ICU.cpp
Debug build of ChakraCore received SIGILL with an ASSERTION FAILURE.

```
"friedrichstra\xDFe 14".toUpperCase(0xa >>> 0.0, "Sticky = true, RegExp.match() result");
```
$ ./ch 99-ef6afe1a505bf421c1e66223554e9ac60503ea96-1586.js 
ASSERTION 18040: (/data/ChakraCore/lib/Runtime/PlatformAgnostic/Platform/Linux/UnicodeText.ICU.cpp, line 401) error == ApiError::NoError
 Failure: (error == ApiError::NoError)
Illegal instruction (core dumped)

$ file ch
ch: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32

OS: Ubuntu 16.04.3 LTS
Arch: x86_64

```
ASSERTION 18047: (/data/ChakraCore/lib/Runtime/PlatformAgnostic/Platform/Linux/UnicodeText.ICU.cpp, line 401) error == ApiError::NoError
 Failure: (error == ApiError::NoError)

Thread 1 "ch" received signal SIGILL, Illegal instruction.
0x00007ffff610e51d in PlatformAgnostic::UnicodeText::ChangeStringCaseInPlace (
    caseFlags=PlatformAgnostic::UnicodeText::CaseFlagsUpper, stringToChange=0x7ffff7e35440 u"FRIEDRICHSTRASSE 1췍", 
    bufferLength=18) at /data/ChakraCore/lib/Runtime/PlatformAgnostic/Platform/Linux/UnicodeText.ICU.cpp:401
401	            Assert(error == ApiError::NoError);
(gdb) bt
#0  0x00007ffff610e51d in PlatformAgnostic::UnicodeText::ChangeStringCaseInPlace (
    caseFlags=PlatformAgnostic::UnicodeText::CaseFlagsUpper, stringToChange=0x7ffff7e35440 u"FRIEDRICHSTRASSE 1췍", 
    bufferLength=18) at /data/ChakraCore/lib/Runtime/PlatformAgnostic/Platform/Linux/UnicodeText.ICU.cpp:401
#1  0x00007ffff5ef8335 in Js::JavascriptString::ToCaseCore (pThis=0x7ff7f0538c60, toCase=Js::JavascriptString::ToUpper)
    at /data/ChakraCore/lib/Runtime/Library/JavascriptString.cpp:2306
#2  0x00007ffff5ef9134 in Js::JavascriptString::EntryToUpperCase (function=0x7ff7f0424140, callInfo=...)
    at /data/ChakraCore/lib/Runtime/Library/JavascriptString.cpp:2235
#3  0x00007ffff5fb73ce in amd64_CallFunction () at /data/ChakraCore/lib/Runtime/Library/amd64/JavascriptFunctionA.S:100
#4  0x00007ffff5d1fed5 in Js::JavascriptFunction::CallFunction<true> (function=0x7ff7f0424140, 
    entryPoint=0x7ffff5ef8db0 <Js::JavascriptString::EntryToUpperCase(Js::RecyclableObject*, Js::CallInfo, ...)>, 
    args=..., useLargeArgCount=false) at /data/ChakraCore/lib/Runtime/Library/JavascriptFunction.cpp:1312
#5  0x00007ffff5b2f24e in Js::InterpreterStackFrame::OP_CallCommon<Js::OpLayoutDynamicProfile<Js::OpLayoutT_CallIWithICIndex<Js::LayoutSizePolicy<(Js::LayoutSize)0> > > > (this=0x7fffffffc110, playout=0x7ffff7e38320, function=0x7ff7f0424140, 
    flags=2, spreadIndices=0x0) at /data/ChakraCore/lib/Runtime/Language/InterpreterStackFrame.cpp:3918
#6  0x00007ffff5b2edfb in Js::InterpreterStackFrame::OP_ProfileCallCommon<Js::OpLayoutDynamicProfile<Js::OpLayoutT_CallIWithICIndex<Js::LayoutSizePolicy<(Js::LayoutSize)0> > > > (this=0x7fffffffc110, playout=0x7ffff7e38320, 
    function=0x7ff7f0424140, flags=0, profileId=0, inlineCacheIndex=0, spreadIndices=0x0)
    at /data/ChakraCore/lib/Runtime/Language/InterpreterStackFrame.cpp:3942
#7  0x00007ffff5b01d48 in Js::InterpreterStackFrame::OP_ProfiledCallIWithICIndex<Js::OpLayoutT_CallIWithICIndex<Js::LayoutSizePolicy<(Js::LayoutSize)0> > > (this=0x7fffffffc110, playout=0x7ffff7e38320)
    at /data/ChakraCore/lib/Runtime/./Language/InterpreterStackFrame.h:503
#8  0x00007ffff59bc2e9 in Js::InterpreterStackFrame::ProcessProfiled (this=0x7fffffffc110)
    at /data/ChakraCore/lib/Runtime/Language/InterpreterHandler.inl:87
#9  0x00007ffff5950f5d in Js::InterpreterStackFrame::Process (this=0x7fffffffc110)
    at /data/ChakraCore/lib/Runtime/Language/InterpreterStackFrame.cpp:3415
#10 0x00007ffff5950186 in Js::InterpreterStackFrame::InterpreterHelper (function=0x7ff7f04c4240, args=..., 
    returnAddress=0x7ff7f0440fa2, addressOfReturnAddress=0x7fffffffc7b8, asmJsReturn=0x0)
    at /data/ChakraCore/lib/Runtime/Language/InterpreterStackFrame.cpp:2002
#11 0x00007ffff594ee4b in Js::InterpreterStackFrame::InterpreterThunk (layout=0x7fffffffc7d0)
    at /data/ChakraCore/lib/Runtime/Language/InterpreterStackFrame.cpp:1733
#12 0x00007ff7f0440fa2 in ?? ()
#13 0x00007fffffffc7e0 in ?? ()
#14 0x00007ffff5fb73ce in amd64_CallFunction () at /data/ChakraCore/lib/Runtime/Library/amd64/JavascriptFunctionA.S:100
Backtrace stopped: frame did not save the PC
```

### PR DESCRIPTION
**Pull request by jackhorton**
Fixes #421 
Fixes #4526 

Also fixes a previously untracked? bug where String.prototype.toLocale{Upper|Lower}Case.call(null) would print "null" rather than throwing a TypeError.

### CODE REVIEW

File: lib/Runtime/Library/JavascriptString.cpp [hunk 1]
```diff
@@ -2216,113 +2214,52 @@ namespace Js
...
+        charcount_t requiredStringLength = 0;
 
-        // Try to find out the chars that do not need casing (in the ASCII range)
-        if (toCase == ToUpper)
+        if (pThis->GetLength() == 1)
         {
-            while (i < inStrLim)
+            // Fast path for one character strings
+            char16 inChar = pThis->GetSz()[0];
+            char16 oneCharAttempt[2] = { inChar, 0 };
```
Comment (MSLaguana) [inline lib/Runtime/Library/JavascriptString.cpp]: What is the point of this buffer? Does the `TryChangeStringLinguisticCaseInPlace` function expect a null terminated string rather than a length-specified string?

File: lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.ICU.cpp [hunk 2]
```diff
@@ -224,55 +227,44 @@ namespace PlatformAgnostic
...
+                    (UChar*) sourceString, sourceLength, locale, &errorCode);
             }
 
-            if (U_FAILURE(errorCode) &&
-                !(destLength == 0 && errorCode == U_BUFFER_OVERFLOW_ERROR))
-            {
-                *pErrorOut = TranslateUErrorCode(errorCode);
-                return -1;
-            }
+            AssertMsg(resultStringLength > 0, "u_strToCase must return required destString length");
```
Comment (MSLaguana) [inline lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.ICU.cpp]: Is this assert hit if you try to change the case of the empty string?

Comment (MSLaguana) [reply lib/Runtime/PlatformAgnostic/Platform/Common/UnicodeText.ICU.cpp]: Oh nevermind, I see there's a check earlier on that the source length is greater than 0; I'm not sure I see where the caller enforces that though

File: lib/Runtime/Library/JavascriptString.cpp [hunk 3]
```diff
@@ -2216,113 +2214,42 @@ namespace Js
...
-        if (i == inStrLim) { return pThis; }
+        ScriptContext* scriptContext = pThis->type->GetScriptContext();
 
-        // Otherwise, copy the string and start casing
-        charcount_t countToCase = (charcount_t)(inStrLim - i);
-        BufferStringBuilder builder(count, pThis->type->GetScriptContext());
-        char16 *outStr = builder.DangerousGetWritableBuffer();
+        ApiError error = ApiError::NoError;
+        // pre-flight to get the length required, as it may be longer than the original string
+        charcount_t requiredStringLength = ChangeStringLinguisticCase<toUpper, useInvariant>(pThis->GetSz(), pThis->GetLength(), nullptr, 0, &error);
```
Comment (jackhorton) [inline lib/Runtime/Library/JavascriptString.cpp]: So some API implementation details, here... ICU's version of this API allows you to pass in a non-null buffer for pre-flighting, in case you happen to provide one long enough. Intl takes advantage of this, where we will basically try a function once with a buffer of a guessed length, and if that fails, allocate a new buffer using the actual length returned from the first attempt. We could do that here, where our initial guess could be to assume that the cased string will be as long as the non-cased string, which will be true in lots of cases. However, the Windows/no-ICU version of this function doesn't support that functionality, because if the buffer isn't long enough, it returns a required length of 0 to indicate failure. So, to be safe, we always pass nullptr here. Its one thing to implement this functionality here separately for ICU and no-ICU, but does that make the ChangeStringLinguisticCase API too complicated?

Comment (dilijev) [reply lib/Runtime/Library/JavascriptString.cpp]: I'm tempted to say open another issue to track this as a possible optimization but for now it's fine.

File: lib/Runtime/Library/JavascriptString.cpp [hunk 4]
```diff
@@ -2137,9 +2140,20 @@ namespace Js
...
 
+    template<bool toUpper>
+    JavascriptString* JavascriptString::ToLocaleCaseHelper(JavascriptString* pThis)
+    {
+        // TODO: implement locale-sensitive Intl versions of these functions
+        return ToCaseCore<toUpper, false>(pThis);
+    }
+
+
     Var JavascriptString::EntryToLowerCase(RecyclableObject* function, CallInfo callInfo, ...)
```
Comment (dilijev) [inline lib/Runtime/Library/JavascriptString.cpp]: nit: extra blank line above

File: lib/Runtime/Library/JavascriptString.cpp [hunk 5]
```diff
@@ -2125,7 +2125,10 @@ namespace Js
 
         Assert(!(callInfo.Flags & CallFlags_New));
 
-        return ToLocaleCaseHelper(args[0], false, scriptContext);
+        JavascriptString * pThis = nullptr;
+        GetThisStringArgument(args, scriptContext, _u("String.prototype.toLocaleUpperCase"), &pThis);
```
Comment (dilijev) [inline lib/Runtime/Library/JavascriptString.cpp]: >toLocaleUpperCase [](start = 72, length = 17)

These labels are mixed up between the two functions.

File: lib/Runtime/Library/JavascriptString.cpp [hunk 6]
```diff
@@ -2137,9 +2140,20 @@ namespace Js
...
     }
 
+    template<bool toUpper>
+    JavascriptString* JavascriptString::ToLocaleCaseHelper(JavascriptString* pThis)
+    {
+        // TODO: implement locale-sensitive Intl versions of these functions
+        return ToCaseCore<toUpper, false>(pThis);
+    }
+
+
```
Comment (dilijev) [inline lib/Runtime/Library/JavascriptString.cpp]: nit: extra blank line below this function

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp [hunk 7]
```diff
@@ -46,7 +46,8 @@ namespace PlatformAgnostic
             return true;
         }
 
-        int32 ChangeStringLinguisticCase(CaseFlags caseFlags, const char16* sourceString, uint32 sourceLength, char16* destString, uint32 destLength, ApiError* pErrorOut)
+        template<bool toUpper, bool useInvariant>
+        charcount_t ChangeStringLinguisticCase(const char16* sourceString, charcount_t sourceLength, char16* destString, charcount_t destLength, ApiError* pErrorOut)
```
Comment (dilijev) [inline lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: >Linguistic [](start = 32, length = 10)

how do we handle the langtag for this?

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp [hunk 8]
```diff
@@ -120,30 +121,6 @@ namespace PlatformAgnostic
             }
         }
 
-        uint32 ChangeStringCaseInPlace(CaseFlags caseFlags, char16* stringToChange, uint32 bufferLength)
```
Comment (dilijev) [inline lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: >ChangeStringCaseInPlace [](start = 15, length = 23)

is this not needed anymore? I think this was an optimization for ASCII strings.

File: test/Strings/unicode_toUpperCase_toLowerCase.js [hunk 9]
```diff
@@ -6,327 +6,345 @@
...
+            assert.areEqual("\u1FBA\u0399", "\u1FB2".toUpperCase(), "Expecting Greek lower-case alpha with varia and ypogegrammeni");
+            assert.areEqual("\u0386\u0399", "\u1FB4".toUpperCase(), "Expecting Greek lower-case alpha with oxia and ypogegrammeni");
+            assert.areEqual("\u1FCA\u0399", "\u1FC2".toUpperCase(), "Expecting Greek lower-case eta with varia and ypogegrammeni");
+            assert.areEqual("\u0389\u0399", "\u1FC4".toUpperCase(), "Expecting Greek lower-case eta with oxia and ypogegrammeni");
+            assert.areEqual("\u1FFA\u0399", "\u1FF2".toUpperCase(), "Expecting Greek lower-case omega with varia and ypogegrammeni");
+            assert.areEqual("\u038F\u0399", "\u1FF4".toUpperCase(), "Expecting Greek lower-case omega with oxia and ypogegrammeni");
+            assert.areEqual("\u0391\u0342\u0399", "\u1FB7".toUpperCase(), "Expecting Greek lower-case alpha with perispomeni and ypogegrammeni");
+            assert.areEqual("\u0397\u0342\u0399", "\u1FC7".toUpperCase(), "Expecting Greek lower-case eta with perispomeni and ypogegrammeni");
+            assert.areEqual("\u03A9\u0342\u0399", "\u1FF7".toUp
```
Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: >} [](start = 8, length = 1)

Should add a test for the proper error thrown for the `String.prototype.toLocale{Upper|Lower}Case.call(null)` case

Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: >upper [](start = 103, length = 5)

lower-case?

Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: >capital [](start = 139, length = 7)

again lower case because you're using .toLowerCase?

Probably just check over these comments again.

File: test/Strings/unicode_toUpperCase_toLowerCase.js [hunk 10]
```diff
@@ -6,327 +6,345 @@
...
+    },
+    {
+        name: "Special casing toUpperCase",
+        body: function () {
+            assert.areEqual("\u0130", "\u0130".toUpperCase(), "Expecting Latin upper-case i with dot above");
+            if (WScript.Platform.INTL_LIBRARY !== "icu") {
+                // Win32's version of toUpperCase does not support growing strings in the uppercase operation (see CharUpperBuff)
+                return;
+            }
+            assert.areEqual("\u0053\u0053", "\u00DF".toUpperCase(), "Expecting Latin lower-case sharp s");
```
Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: >lower-case [](start = 85, length = 10)

you used toUpperCase so this comment should reflect that

File: test/Strings/unicode_toUpperCase_toLowerCase.js [hunk 11]
```diff
@@ -6,327 +6,345 @@
...
+            assert.areEqual("\u1F09\u0399", "\u1F81".toUpperCase(), "Expecting Greek lower-case alpha with dasia and ypogegrammeni");
+            assert.areEqual("\u1F0A\u0399", "\u1F82".toUpperCase(), "Expecting Greek lower-case alpha with psili and varia and ypogegrammeni");
+            assert.areEqual("\u1F0B\u0399", "\u1F83".toUpperCase(), "Expecting Greek lower-case alpha with dasia and varia and ypogegrammeni");
+            assert.areEqual("\u1F0C\u0399", "\u1F84".toUpperCase(), "Expecting Greek lower-case alpha with psili and oxia and ypogegrammeni");
+            assert.areEqual("\u1F0D\u0399", "\u1F85".toUpperCase(), "Expecting Greek lower-case alpha with dasia and oxia and ypogegrammeni");
+            assert.areEqual("\u1F0E\u0399", "\u1F86".toUpperCase(), "Expecting Greek lower-case alpha with psili and perispomeni and ypogegrammeni");
+            assert.areEqual("\u1F0F\u0399", "\u1F87".toUpperCase(), "Expecting Greek lower-case alpha with dasia and perispomeni and ypogegrammeni");
+            assert.areEqual("\u1F08\u0399", "\u1F88".toUpperCase(), "Expecting Greek upper-case alpha with psili and prosgegrammeni");
```
Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: some mixing of lowercase and uppercase but all operations are toUpperCase

File: test/Strings/unicode_toUpperCase_toLowerCase.js [hunk 12]
```diff
@@ -6,327 +6,345 @@
...
-        assert.areEqual("a\u03C3\u00ADb", "A\u03A3\u00ADB".toLowerCase(), "Expecting sigma followed by soft hyphen (00ad)");
-        assert.areEqual("a\u03C3\uD834\uDE42b", "A\u03A3\uD834\uDE42B".toLowerCase(), "Expecting sigma followed by combining Greek musical triseme (d834 de42 = 1d242)");
-        //assert.areEqual("a\u03C2\u0345", "A\u03A3\u0345".toLowerCase(), "Expecting sigma followed by combining Greek ypogegrammeni (0345)");  // not yet implemented
-        assert.areEqual("a\u03C3\u0345\u03B1", "A\u03A3\u0345\u0391".toLowerCase(), "Expecting sigma followed by combining Greek ypogegrammeni (0345), Greek upper-case alpha (0391)");
-    }
-  },
+    {
+        name: "Edge cases",
+        body() {
+            assert.areEqual("", "".toUpperCase(), "Empty string should return itself (toUpperCase)");
```
Comment (dilijev) [inline test/Strings/unicode_toUpperCase_toLowerCase.js]: >toUpperCase [](start = 35, length = 11)

Did you add any tests for toLocale{Upper|Lower}Case ?

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp (hunk 7 above)
Comment (jackhorton) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: POSIX doesn't have any idea of locale for this implementation, but this is only used in the --no-icu case, which as far as I know is anything goes in terms of spec compliance. This didn't handle system vs root locale before, either.

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp (hunk 8 above)
Comment (jackhorton) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: I removed it because we couldn't guarantee in-place conversion for any arbitrary string (any ICU implementation takes a src and dest buffer, and if characters can expand from source to dest, doing conversions in place using the same buffer for both args would produce incorrect results). I suppose I can add this back and fail for non-ASCII?

File: test/Strings/unicode_toUpperCase_toLowerCase.js (hunk 9 above)
Comment (jackhorton) [reply test/Strings/unicode_toUpperCase_toLowerCase.js]: These were added previously. I could try to go through and add them all again, but I imagine they were auto-generated?

File: test/Strings/unicode_toUpperCase_toLowerCase.js (hunk 12 above)
Comment (jackhorton) [reply test/Strings/unicode_toUpperCase_toLowerCase.js]: I think that should probably go in a different file.

File: lib/Runtime/Library/JavascriptString.cpp (hunk 5 above)
Comment (jackhorton) [reply lib/Runtime/Library/JavascriptString.cpp]: Good catch

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp (hunk 8 above)
Comment (jackhorton) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: I think I am still against this function because in reality, this is only called when converting JavascriptStrings, which are immutable and can't be changed in place regardless. So, unless we can come up with a case where we are casing a string from native code that _isn't_ a JavascriptString, I think this function just adds more complexity.

**Comment by jackhorton** [comment]
@dilijev Unless you feel particularly strongly about it, I feel like I should leave the message issues in the Unicode test as is and open a bug to regenerate that test for Unicode 10 soon.

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp (hunk 8 above)
Comment (dilijev) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: Sounds good to me. If it regresses perf in a scenario we are tracking we can optimize later.

**Comment by dilijev** [comment]
@jackhorton that's fine if there's a test to track. Keeping in mind the test is a bit misleading at the moment (especially if something fails could cause confusion).

We both have Unicode 10 update work to do :P

File: lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp (hunk 8 above)
Comment (jackhorton) [reply lib/Runtime/PlatformAgnostic/Platform/POSIX/UnicodeText.cpp]: Is it worth submitting an IB run with this change?
